snap/snapcraft.yaml

```diff
@@ -0,0 +1,147 @@
+name: livebook
+base: core20
+version: "0.6.3"
+summary: Interactive & collaborative code notebooks in Elixir
+description: |
+  Livebook is a web application for writing interactive and collaborative code notebooks for Elixir.
+
+grade: devel
+confinement: strict
+architectures:
+  - build-on: amd64
+  - build-on: arm64
+
+package-repositories:
+  - type: apt
+    url: http://binaries.erlang-solutions.com/debian
+    key-id: "434975BD900CCBE4F7EE1B1ED208507CA14F4FCA"
+    suites:
+      - "focal"
+    components:
+      - "contrib"
+    architectures:
+      - "amd64"
+      - "arm64"
+
+hooks:
+  configure:
+    plugs: [network, network-bind]
+
+apps:
+  livebook:
+    extensions: [gnome-3-38]
+    command: bin/livebook-indicator
+    plugs:
+      - desktop
+      - gsettings
+      - wayland
+      - x11
+
+  server:
+    command: wrappers/start-livebook.sh
+    plugs:
+      - network-bind
+      - network
+      - home
+    environment:
+      "LIVEBOOK_HOME": $SNAP_REAL_HOME
+      "LIVEBOOK_DATA_PATH": $SNAP_USER_COMMON
+
+layout:
+  /usr/lib/erlang:
+    symlink: $SNAP/usr/lib/erlang
+  /usr/bin/git:
+    symlink: $SNAP/usr/bin/git
+
+parts:
+  wrappers:
+    plugin: dump
+    source: snap/local/wrappers
+    organize:
+      "*": "wrappers/"
+
+  elixir:
+    plugin: nil
+    source: https://github.com/elixir-lang/elixir
+    source-type: git
+    source-tag: v1.14.0-rc.1
+
+    override-build: |
+      PREFIX=$SNAPCRAFT_PART_INSTALL make install
+
+    build-packages:
+      - esl-erlang
+
+    stage-packages:
+      - esl-erlang
+      - libodbc1
+      - libglu1-mesa
+      - libwxgtk-webview3.0-gtk3-0v5
+
+  livebook:
+    after:
+      - elixir
+    source: https://github.com/livebook-dev/livebook
+    source-type: git
+    source-tag: v0.6.3
+    plugin: nil
+    override-build: |
+      mix local.hex --force && mix local.rebar --force && \
+      mix do deps.get, deps.compile && \
+      MIX_ENV=prod mix do compile, release livebook && \
+      mkdir -p $SNAPCRAFT_PART_INSTALL && \
+      cp -r _build/prod/rel/livebook $SNAPCRAFT_PART_INSTALL/livebook && \
+      echo $(cat /dev/urandom | env LC_ALL=C tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1) > $SNAPCRAFT_PART_INSTALL/livebook/releases/COOKIE
+    stage:
+      - livebook
+    build-packages:
+      - build-essential
+      - gnupg2
+    stage-packages:
+      - libcrypt1
+      - libtinfo6
+      # The following staged dependencies are inspired by the livebook Dockerfile
+      # ... as are the comments.
+      # Runtime dependencies
+      - build-essential
+      - ca-certificates
+      - libncurses5-dev
+      - git # In case someone uses `Mix.install/2` and point to a git repo
+      - wget # Additional standard tools
+      - cmake # In case someone uses Torchx for Nx
+
+  indicator:
+    after:
+      - livebook
+    source: https://github.com/mz2/livebook-indicator
```

Comment:
> Besides the GUI part, the system tray, we need two other things from the OS, handling the "open file" and "open url" events

Yep, I think this should be pretty simple. I can investigate this on top of the current indicator solution if the Rust / GTK route seems sane to you.

> it sounds like all we need to expose is a public git repo which snapcraft will pull from. So it seems it is fundamentally different model than we have with app bundler?

To be precise, snapcraft _allows for_ a fundamentally different model but does not mandate it. I suppose this `source-type: git` mode of fetching the source is motivated by the common need to support the publisher / packager and upstream being different entities. It's also possible to refer to the source locally from the same repository instead of using git (by changing `source-type: local` in `snapcraft.yaml` in the `livebook` part), and then `snapcraft` would at build time instead use the directory containing the `snap` directory (i.e. the root of the repo in this case) as the root source directory that is made available to the build process.

I made a separate indicator app instead of investigating Linux enabling the app bundler route mostly since a) on a cursory look the wxgtk system tray support is for the legacy system tray protocol that libappindicator supersedes and b) I'm no expert on the subject and the libappindicator / gtk + Rust route involved less abstractions between me and making an indicator dingus show up :smile:

Reply:
> I can investigate this on top of the current indicator solution if the Rust / GTK route seems sane to you.

We'd prefer not to take the path of separate Linux-specific codebase for parts of the functionality. All the caveats aside, wx is still appealing to us exactly because we have a single tech stack that abstracts OS-specific details. Something we might consider in the future is going all in on something like Tauri (or putting together some of its underlying components or implementing more things ourselves in Rust or others)

Personally I think https://github.com/mz2/livebook-indicator is really cool and I think there could be many more apps using it - all they need is a system tray indicator that controls a server that runs in the background. For Livebook specifically I'd personally consider dropping our wx based solution in favour of this if it would support Mac & Windows too.

All that being said just letting you know that we're also investigating WebView-based solutions (so that we could have an iPad app) so we might be taking a very different approach in the future after all.